src/viam/ur/module/ur_arm.cpp

```diff
@@ -172,128 +184,1284 @@ void write_waypoints_to_csv(const std::string& filepath, const std::list<Eigen::
+            try {
+                if (!arm_conn_->driver->sendRobotProgram()) {
+                    VIAM_SDK_LOG(warn) << "While in independent state, could not send program to robot via driver; disconnecting";
+                    return event_connection_lost_{};
+                }
+            } catch (...) {
+                VIAM_SDK_LOG(warn) << "While in independent state, failed sending program to robot via driver; disconnecting";
+                return event_connection_lost_{};
+            }
         }
 
-        void write_joint_data(vector6d_t& position, vector6d_t& velocity) {
-            ::write_joint_data(position, velocity, arm_joint_positions_stream, unix_time_iso8601(), arm_joint_positions_sample++);
+        // TODO: Grace period?
+        if (!arm_conn_->program_running_flag.load(std::memory_order_acquire)) {
+            VIAM_SDK_LOG(info) << "While in independent state, waiting for callback to toggle program state to running";
+            return std::nullopt;
         }
 
-        std::vector<trajectory_sample_point> samples;
-        std::ofstream arm_joint_positions_stream;
-        std::size_t arm_joint_positions_sample{0};
-        std::promise<void> completion;
+        return event_remote_mode_restored_{};
+    }
 
-        struct cancellation_request {
-            // This constructor needs to be written this way for
-            // std::optional::emplace with no arguments to work.
-            cancellation_request() {}
+    return std::nullopt;
+}
 
-            std::promise<void> promise;
-            std::shared_future<void> future;
-            bool issued{false};
-        };
+std::optional<URArm::state_::event_variant_> URArm::state_::state_independent_::handle_move_request(state_& state) {
+    if (state.move_request_) {
+        std::exchange(state.move_request_, {})->complete_error([this]() -> std::string {
+            switch (reason_) {
+                case reason::k_estopped: {
+                    return "arm is estopped";
+                }
+                case reason::k_local_mode: {
+                    return "arm is in local mode";
+                }
+                case reason::k_both: {
+                    return "arm is in local mode and estopped";
+                }
+                default: {
+                    return "arm is in an unknown and uncontrolled state";
+                }
+            }
+        }());
+    }
+    return std::nullopt;
+}
 
-        std::optional<cancellation_request> cancellation_request;
-    };
-    std::optional<move_request> move_request;
+bool URArm::state_::state_independent_::estopped() const {
+    return reason_ != reason::k_local_mode;
+}
+
+bool URArm::state_::state_independent_::local_mode() const {
+    return reason_ != reason::k_estopped;
+}
+
+URArm::state_::state_variant_ URArm::state_::state_independent_::handle_event(event_estop_cleared_) {
```

Comment:
Fairly boring event handlers from here. They just decide what the next state is given the current state, and instantiate a new state object.